src/BaseMailer.php

```diff
@@ -0,0 +1,179 @@
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var LoggerInterface $logger.
+     */
+    private $logger;
+
+    /**
+     * @param MessageFactoryInterface $messageFactory
+     * @param Composer $composer
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param LoggerInterface $logger
+     */
+    public function __construct(MessageFactoryInterface $messageFactory, Composer $composer, EventDispatcherInterface $eventDispatcher, LoggerInterface $logger)
+    {
+        $this->messageFactory = $messageFactory;
+        $this->composer = $composer;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @return Composer message composer instance.
+     */
+    public function getComposer(): Composer
+    {
+        return $this->composer;
+    }
+
+    /**
+     * Creates a new message instance and optionally composes its body content via view rendering.
+     *
+     * @param string|array|null $view the view to be used for rendering the message body. This can be:
+     *
+     * - a string, which represents the view name for rendering the HTML body of the email.
+     *   In this case, the text body will be generated by applying `strip_tags()` to the HTML body.
+     * - an array with 'html' and/or 'text' elements. The 'html' element refers to the view name
+     *   for rendering the HTML body, while 'text' element is for rendering the text body. For example,
+     *   `['html' => 'contact-html', 'text' => 'contact-text']`.
+     * - null, meaning the message instance will be returned without body content.
+     *
+     * The view to be rendered can be specified in one of the following formats:
+     *
+     * - a relative view name (e.g. "contact") located under [[viewPath]].
+     *
+     * @param array $parameters the parameters (name-value pairs) that will be extracted and made available in the view file.
+     * @return MessageInterface message instance.
+     */
+    public function compose($view = null, array $parameters = []): MessageInterface
+    {
+        $message = $this->createMessage();
+        if ($view === null) {
+            return $message;
+        }
+
+        $this->composer->compose($message, $view, $parameters);
+
+        return $message;
+    }
+
+    /**
+     * Creates a new message instance.
+     * @return MessageInterface message instance.
+     */
+    protected function createMessage(): MessageInterface
+    {
+        return $this->messageFactory->create($this);
+    }
+
+    /**
+     * Sends the given email message.
+     * This method will log a message about the email being sent. 
+     * Child classes should implement [[sendMessage()]] with the actual email sending logic.
+     * @param MessageInterface $message email message instance to be sent
+     * @throws \Throwable throws an exception on send fails.
+     */
+    public function send(MessageInterface $message): void
```

Comment:
Just a reminder, may be dealt with in a subsequent PR: I remember an issue asking for send() to return the number of messages sent (like swiftmailer send() does), or at least a boolean.

Reply:
Are you talking about this issue yiisoft/yii2-swiftmailer#59?

Reply:
For this case, I've a proposal:
1. `send` return bool instead throwing exception, error could be retrieved be `MessageInterface::getError`:
```php
if (!$mailer->send($message)) {
    $error = $message->getError();
    // handling error
}
```

2. `sendMultiple` accepts two parameters, `messages` and `failedMessages`(passed by reference), and return `true` if all messages were sent successfully:
```php
if (!$mailer->send($messages, $failedMessages)) {
    foreach ($failedMessages as $message) {
        // handling error
    }
}
```

@samdark @machour what do you think?

Reply:
@razonyang yes, that's the issue I was talking about.
Your proposed solution looks good to me and a PR would be very welcome!